File: scala/lemkit-model/src/main/scala/io/people8/classify/LinearClassifier.scala
package io.people8.classify

import scala.io.Source
import scala.collection.mutable

import java.io._

import net.liftweb.{ json => liftjson }
import net.liftweb.json.JsonDSL._

import data._
import util.NumNormalizer._

/**
 * A Scala-usable version of a linear classifier.
 *
 * FIXME: Generalize label type
 *
 * @tparam I type of data instance
 * @param parameters Array of weight arrays, one per label; should have
 *   same number of weights per array
 * @param lmap Map from label names to corresponding identifying integers
 * @param fmap Map from feature names to corresponding identifying integers
 * @param featurizer Object to generate features for data instance
 */
@SerialVersionUID(1)
class LinearClassifier[I](
    private val indexer: ClassifierIndexer[I],
    private val parameters: Array[Array[Double]]) extends FeaturizingClassifier[I](indexer) {
  def rawEvalFeatures(feats: Seq[FeatureObservation[Int]]): Seq[Double] = {
    val numClasses = indexer.lmap.size

    // Start with zero array
    val scoresLogs = Array.fill(numClasses)(0.0)

    // Update the scoresLogs based on the features.
    for (fobs <- feats; classIndex <- 0 until numClasses)
      scoresLogs(classIndex) += parameters(classIndex)(fobs.feature) * fobs.magnitude

    scoresLogs.toSeq
  }

  def scores(text: I): Seq[(String, Double)] = {
    labelIndex zip sumNormalize(exp(evalRaw(text)))
  }
}

class InvalidFormatException(msg: String) extends Exception(msg)

/**
 * General utilities for dealing with linear classifiers.
 */
object LinearClassifier {
  val magicNumber = 0x6A48B9DD
  val majorVersion = 1
  val minorVersion = 0
  val labelID = 100
  val featureTypeID = 110
  val featureTypeExact = 1
  val featureTypeHashed = 2
  val featureTypeBloomFilterHashed = 3
  val featureFeaturesID = 111
  val featureMaxFeatsID = 112
  val featureBloomFilterID = 113
  val weightsTypeID = 120
  val weightsTypeDense = 1
  val weightsTypeSparse = 2
  val weightsValuesID = 121

  def apply[I](indexer: ClassifierIndexer[I], params: Array[Array[Double]]) =
    new LinearClassifier(indexer, params)

  def readJSONModel[I](file: String, featurizer: Featurizer[I, String]) =
    readJSONModelSource(Source.fromFile(file), featurizer)

  def readJSONModelSource[I](source: Source,
    featurizer: Featurizer[I, String]) = {
    val filecontents = source.getLines.toSeq.mkString("\n")
    val json = liftjson.parse(filecontents)
    val jsonfeats = json \ "features"
    val fmap = (jsonfeats \ "type").values.toString match {
      case "exact" =>
        new ExactFeatureMap((jsonfeats \ "features").values.
          asInstanceOf[Map[String, BigInt]].map {
            case (key, value) => (key, value.toInt)
          })
      case "hashed" =>
        HashedFeatureMap((jsonfeats \ "maxfeats").values.toString.toInt)
    }
    val lmap = (json \ "labels").values.asInstanceOf[Map[String, BigInt]].map {
      case (key, value) => (key, value.toInt)
    }

    val jsonweights = json \ "weights"
    val weights = (jsonweights \ "type").values.toString match {
      case "dense" =>
        (jsonweights \ "values").values.asInstanceOf[Seq[Seq[Double]]]
      case "sparse" =>
        ???
    }
    val params = weights.map(_.toArray).toArray
    apply(new ClassifierIndexer(lmap, fmap, featurizer), params)
  }

  import java.nio.charset.Charset
  val utf8 = Charset.forName("UTF-8")

  def writeString(out: DataOutput, str: String) {
    val bytes = str.getBytes(utf8)
    out.writeInt(bytes.length)
    out.write(bytes, 0, bytes.length)
  }

  def writeStringSeq(out: DataOutput, labels: Seq[String]) {
    out.writeInt(labels.length)
    labels.foreach(label => writeString(out, label))
  }

  def writeTable(out: DataOutput, table: Map[String, Int]) {
    out.writeInt(table.size)
    for ((label, value) <- table) {
      writeString(out, label)
      out.writeInt(value)
    }
  }

  def readString(in: DataInput) = {
    val len = in.readInt
    val bytes = new Array[Byte](len)
    in.readFully(bytes)
    new String(bytes, utf8)
  }

  def readStringSeq(in: DataInput) = {
    val length = in.readInt
    (for (i <- 0 until length) yield readString(in)).toSeq
  }

  def readTable(in: DataInput) = {
    val size = in.readInt
    val map = mutable.Map[String, Int]()
    for (i <- 0 until size) {
      val label = readString(in)
      val value = in.readInt
      map(label) = value
    }
    map.toMap
  }

  def checkInt(in: DataInput, value: Int) {
    val checked = in.readInt
    if (checked != value)
      throw new InvalidFormatException(
        "Invalid binary format; expected int %s but saw %s" format (
          value, checked))
  }

  def checkShort(in: DataInput, value: Int) {
    val checked = in.readShort
    if (checked != value)
      throw new InvalidFormatException(
        "Invalid binary format; expected short %s but saw %s" format (
          value, checked))
  }

  def readBinaryModel[I](file: String, featurizer: Featurizer[I, String]) = {
    val in = new DataInputStream(new BufferedInputStream(new FileInputStream(file)))
    try {
      readBinaryModelStream(in, featurizer)
    } finally {
      in.close()
    }
  }

  def readBinaryModelStream[I](in: DataInputStream, featurizer: Featurizer[I, String]) = {
    checkInt(in, magicNumber)
    checkShort(in, majorVersion)
    checkShort(in, minorVersion)
    checkShort(in, labelID)
    val lmap = readTable(in)
    checkShort(in, featureTypeID)
    val featureType = in.readShort
    val fmap = if (featureType == featureTypeExact) {
      checkShort(in, featureFeaturesID)
      ExactFeatureMap(readStringSeq(in))
    } else if (featureType == featureTypeHashed) {
      checkShort(in, featureMaxFeatsID)
      HashedFeatureMap(in.readInt)
    } else {
      throw new InvalidFormatException(
        "Invalid binary format; unknown feature map type %s" format
          featureType)
    }
    checkShort(in, weightsTypeID)
    // FIXME: Implement sparse weights
    checkShort(in, weightsTypeDense)
    checkShort(in, weightsValuesID)
    val numClasses = in.readInt
    val parameters = new Array[Array[Double]](numClasses)
    for (i <- 0 until numClasses) {
      val len = in.readInt
      val oneparams = new Array[Double](len)
      for (j <- 0 until len) {
        oneparams(j) = in.readDouble
      }
      parameters(i) = oneparams
    }
    apply(new ClassifierIndexer(lmap, fmap, featurizer), parameters)
  }

  def writeJSONModel[I](classifier: LinearClassifier[I], file: String) = {
    val features = classifier.indexer.fmap match {
      case f: ExactFeatureMap =>
        ("type" -> "exact") ~ ("features" -> f.fmap)
      case f: HashedFeatureMap =>
        ("type" -> "hashed") ~ ("maxfeats" -> f.maxNumberOfFeatures)
    }
    val seqparams = classifier.parameters.map(_.toIndexedSeq).toIndexedSeq
    val json =
      ("labels" -> classifier.indexer.lmap) ~ ("features" -> features) ~
        ("weights" ->
          ("type" -> "dense") ~ ("values" -> seqparams)
        )
    val rendered = liftjson.pretty(liftjson.render(json))
    val out = new BufferedWriter(new FileWriter(new File(file)))
    try {
      out.write(rendered + "\n")
    } finally {
      out.close()
    }
  }

  def writeBinaryModel[I](classifier: LinearClassifier[I], file: String) {
    val out = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)))
    try {
      writeBinaryModel(classifier, out)
    } finally {
      out.close()
    }
  }

  def writeBinaryModel[I](classifier: LinearClassifier[I], out: DataOutputStream) {
    out.writeInt(magicNumber)
    out.writeShort(majorVersion)
    out.writeShort(minorVersion)
    out.writeShort(labelID)
    writeTable(out, classifier.indexer.lmap)
    out.writeShort(featureTypeID)
    classifier.indexer.fmap match {
      case f: ExactFeatureMap => {
        out.writeShort(featureTypeExact)
        out.writeShort(featureFeaturesID)
        writeStringSeq(out, f.fmap.toSeq.sortBy(_._2).unzip._1)
      }
      case f: HashedFeatureMap => {
        out.writeShort(featureTypeHashed)
        out.writeShort(featureMaxFeatsID)
        out.writeInt(f.maxNumberOfFeatures)
      }
    }
    out.writeShort(weightsTypeID)
    out.writeShort(weightsTypeDense)
    out.writeShort(weightsValuesID)
    out.writeInt(classifier.parameters.length)
    for (oneparams <- classifier.parameters) {
      out.writeInt(oneparams.length)
      for (param <- oneparams) {
        out.writeDouble(param)
      }
    }
  }

}

File: go/classify/LinearClassifier.go
package classify

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"os"
)

type LinearClassifier struct {
	FeaturizingClassifier
	Parameters [][]float64
}

func (c *LinearClassifier) IntRawScores(feats []IntFeatureObservation) []float64 {
	numClasses := len(c.indexer.lmap)
	scores := make([]float64, numClasses)
	for _, fobs := range feats {
		for class := 0; class < numClasses; class++ {
			scores[class] += c.Parameters[class][fobs.Feature] * fobs.Magnitude
		}
	}
	return scores
}

func (c *LinearClassifier) RawScores(feats []FeatureObservation) []float64 {
	return c.IntRawScores(c.indexer.Index(feats))
}

func (c *LinearClassifier) Scores(feats []FeatureObservation) []float64 {
	scores := c.RawScores(feats)
	for i := range scores {
		scores[i] = math.Exp(scores[i])
	}
	sum := 0.0
	for i := range scores {
		sum += scores[i]
	}
	if sum != 0.0 {
		for i := range scores {
			scores[i] /= sum
		}
	}
	return scores
}

func (c *LinearClassifier) ScoresMap(feats []FeatureObservation) map[string]float64 {
	labels := c.Labels()
	scores := c.Scores(feats)
	if len(labels) != len(scores) {
		panic(fmt.Sprintf("wrong-length sequences: len(labels)=%s, len(scores)=%s",
			len(labels), len(scores)))
	}
	retval := make(map[string]float64, 0)
	for i := range labels {
		retval[labels[i]] = scores[i]
	}
	return retval
}

func (c *LinearClassifier) Predict(feats []FeatureObservation) (label string, score float64) {
	labels := c.Labels()
	scores := c.Scores(feats)
	if len(labels) != len(scores) {
		panic(fmt.Sprintf("wrong-length sequences: len(labels)=%s, len(scores)=%s",
			len(labels), len(scores)))
	}
	label = labels[0]
	score = scores[0]
	for i := 1; i < len(labels); i++ {
		if scores[i] > score {
			score = scores[i]
			label = labels[i]
		}
	}
	return
}

/**
 * General utilities for dealing with linear classifiers.
 */
const (
	magicNumber       = 0x6A48B9DD
	majorVersion      = 1
	minorVersion      = 0
	labelID           = 100
	featureTypeID     = 110
	featureTypeExact  = 1
	featureTypeHashed = 2
	featureFeaturesID = 111
	featureMaxFeatsID = 112
	weightsTypeID     = 120
	weightsTypeDense  = 1
	weightsTypeSparse = 2
	weightsValuesID   = 121
)

type JSONClassifier struct {
	Labels   map[string]int
	Features struct {
		Type     string
		Features map[string]int
		MaxFeats int
	}
	Weights struct {
		Type   string
		Values [][]float64
	}
}

func ReadJSONModel(file string) (*LinearClassifier, error) {
	in, err := os.Open(file)
	if err != nil {
		return nil, err
	}
	return ReadJSONModelReader(in)
}

func ReadJSONModelReader(in io.Reader) (*LinearClassifier, error) {
	dec := json.NewDecoder(in)
	var v JSONClassifier
	if err := dec.Decode(&v); err != nil {
		return nil, err
	}

	if v.Labels == nil {
		return nil, fmt.Errorf("Missing labels in JSON")
	}

	var fmap FeatureMap
	if v.Features.Type == "exact" {
		if v.Features.Features == nil {
			return nil, fmt.Errorf("Missing features -> features in JSON")
		}
		fmap = NewExactFeatureMap(v.Features.Features)
	} else if v.Features.Type == "hashed" {
		if v.Features.MaxFeats <= 0 {
			return nil, fmt.Errorf("Missing or invalid maxfeats in JSON: %s",
				v.Features.MaxFeats)
		}
		fmap = NewHashedFeatureMap(v.Features.MaxFeats)
	} else {
		return nil, fmt.Errorf("Missing or invalid features -> type in JSON: %s",
			v.Features.Type)
	}

	if v.Weights.Type != "dense" {
		return nil, fmt.Errorf("Missing or invalid weights -> type in JSON: %s",
			v.Weights.Type)
	}

	return &LinearClassifier{FeaturizingClassifier{&ClassifierIndexer{v.Labels, fmap}}, v.Weights.Values}, nil
}

func writeInt(out io.Writer, value int) error {
	v := int32(value)
	return binary.Write(out, binary.BigEndian, &v)
}

func writeShort(out io.Writer, value int16) error {
	return binary.Write(out, binary.BigEndian, &value)
}

func writeString(out io.Writer, str string) error {
	err := writeInt(out, len(str))
	if err != nil {
		return err
	}
	_, err = io.WriteString(out, str)
	return err
}

func writeStringSeq(out io.Writer, seq []string) error {
	l := len(seq)
	err := writeInt(out, l)
	if err != nil {
		return err
	}
	for i := 0; i < l; i++ {
		err = writeString(out, seq[i])
		if err != nil {
			return err
		}
	}
	return nil
}

func writeTable(out io.Writer, table map[string]int) error {
	l := len(table)
	err := writeInt(out, l)
	if err != nil {
		return err
	}
	for key := range table {
		value := table[key]
		err = writeString(out, key)
		if err != nil {
			return err
		}
		err = writeInt(out, value)
		if err != nil {
			return err
		}
	}
	return nil
}

func readInt(in io.Reader) (int, error) {
	var value int32
	err := binary.Read(in, binary.BigEndian, &value)
	return int(value), err
}

func readShort(in io.Reader) (int16, error) {
	var value int16
	err := binary.Read(in, binary.BigEndian, &value)
	return value, err
}

func readFloat64(in io.Reader) (float64, error) {
	var value float64
	err := binary.Read(in, binary.BigEndian, &value)
	return value, err
}

func readString(in io.Reader) (string, error) {
	len, err := readInt(in)
	if err != nil {
		return "", err
	}
	bytes := make([]byte, len)
	if _, err := io.ReadFull(in, bytes); err != nil {
		return "", err
	}
	return string(bytes), nil
}

func readStringSeq(in io.Reader) ([]string, error) {
	len, err := readInt(in)
	if err != nil {
		return nil, err
	}
	ret := make([]string, len)
	for i := 0; i < len; i++ {
		str, err := readString(in)
		if err != nil {
			return nil, err
		}
		ret[i] = str
	}
	return ret, nil
}

func readTable(in io.Reader) (map[string]int, error) {
	len, err := readInt(in)
	if err != nil {
		return nil, err
	}
	ret := make(map[string]int, len)
	for i := 0; i < len; i++ {
		key, err := readString(in)
		if err != nil {
			return nil, err
		}
		value, err := readInt(in)
		if err != nil {
			return nil, err
		}
		ret[key] = value
	}
	return ret, nil
}

func checkInt(in io.Reader, value int) error {
	checked, err := readInt(in)
	if err != nil {
		return err
	}
	if checked != value {
		return fmt.Errorf("Invalid binary format; expected int %s but saw %s",
			value, checked)
	}
	return nil
}

func checkShort(in io.Reader, value int16) error {
	checked, err := readShort(in)
	if err != nil {
		return err
	}
	if checked != value {
		return fmt.Errorf("Invalid binary format; expected short %s but saw %s",
			value, checked)
	}
	return nil
}

func ReadBinaryModel(file string) (*LinearClassifier, error) {
	in, err := os.Open(file)
	if err != nil {
		return nil, err
	}
	return ReadBinaryModelReader(in)
}

func ReadBinaryModelReader(in io.Reader) (lc *LinearClassifier, err error) {
	err = checkInt(in, magicNumber)
	if err != nil {
		return
	}
	err = checkShort(in, majorVersion)
	if err != nil {
		return
	}
	err = checkShort(in, minorVersion)
	if err != nil {
		return
	}
	err = checkShort(in, labelID)
	if err != nil {
		return
	}
	var lmap map[string]int
	lmap, err = readTable(in)
	if err != nil {
		return
	}
	err = checkShort(in, featureTypeID)
	if err != nil {
		return
	}
	var featureType int16
	featureType, err = readShort(in)
	if err != nil {
		return
	}
	var fmap FeatureMap
	if featureType == featureTypeExact {
		err = checkShort(in, featureFeaturesID)
		if err != nil {
			return
		}
		var seq []string
		seq, err = readStringSeq(in)
		if err != nil {
			return
		}
		exactmap := make(map[string]int, len(seq))
		for i := 0; i < len(seq); i++ {
			exactmap[seq[i]] = i
		}
		fmap = NewExactFeatureMap(exactmap)
	} else if featureType == featureTypeHashed {
		err = checkShort(in, featureMaxFeatsID)
		if err != nil {
			return
		}
		var maxNumFeatures int
		maxNumFeatures, err = readInt(in)
		if err != nil {
			return
		}
		fmap = NewHashedFeatureMap(maxNumFeatures)
	} else {
		err = fmt.Errorf("Invalid binary format; unknown feature map type %s",
			featureType)
		return
	}
	err = checkShort(in, weightsTypeID)
	if err != nil {
		return
	}
	// FIXME: Implement sparse weights
	err = checkShort(in, weightsTypeDense)
	if err != nil {
		return
	}
	err = checkShort(in, weightsValuesID)
	if err != nil {
		return
	}
	var numClasses int
	numClasses, err = readInt(in)
	if err != nil {
		return
	}
	parameters := make([][]float64, numClasses)
	for i := 0; i < numClasses; i++ {
		var numWeights int
		numWeights, err = readInt(in)
		if err != nil {
			return
		}
		oneparams := make([]float64, numWeights)
		for j := 0; j < numWeights; j++ {
			var weight float64
			weight, err = readFloat64(in)
			if err != nil {
				return
			}
			oneparams[j] = weight
		}
		parameters[i] = oneparams
	}

	lc = &LinearClassifier{FeaturizingClassifier{&ClassifierIndexer{lmap, fmap}}, parameters}
	return
}

//  def writeJSONModel[I](classifier: LinearClassifier[I], file: String) = {
//    val features = classifier.indexer.fmap match {
//      case f: ExactFeatureMap =>
//        ("type" -> "exact") ~ ("features" -> f.fmap)
//      case f: HashedFeatureMap =>
//        ("type" -> "hashed") ~ ("maxfeats" -> f.maxNumberOfFeatures)
//    }
//    val seqparams = classifier.parameters.map(_.toIndexedSeq).toIndexedSeq
//    val json =
//      ("labels" -> classifier.indexer.lmap) ~ ("features" -> features) ~
//        ("weights" ->
//          ("type" -> "dense") ~ ("values" -> seqparams)
//        )
//    val rendered = liftjson.pretty(liftjson.render(json))
//    val out = new BufferedWriter(new FileWriter(new File(file)))
//    try {
//      out.write(rendered + "\n")
//    } finally {
//      out.close()
//    }
//  }
//
//  def writeBinaryModel[I](classifier: LinearClassifier[I], file: String) {
//    val out = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)))
//    try {
//      writeBinaryModel(classifier, out)
//    } finally {
//      out.close()
//    }
//  }
//
//  def writeBinaryModel[I](classifier: LinearClassifier[I], out: DataOutputStream) {
//    out.writeInt(magicNumber)
//    out.writeShort(majorVersion)
//    out.writeShort(minorVersion)
//    out.writeShort(labelID)
//    writeTable(out, classifier.indexer.lmap)
//    out.writeShort(featureTypeID)
//    classifier.indexer.fmap match {
//      case f: ExactFeatureMap => {
//        out.writeShort(featureTypeExact)
//        out.writeShort(featureFeaturesID)
//        writeStringSeq(out, f.fmap.toSeq.sortBy(_._2).unzip._1)
//      }
//      case f: HashedFeatureMap => {
//        out.writeShort(featureTypeHashed)
//        out.writeShort(featureMaxFeatsID)
//        out.writeInt(f.maxNumberOfFeatures)
//      }
//    }
//    out.writeShort(weightsTypeID)
//    out.writeShort(weightsTypeDense)
//    out.writeShort(weightsValuesID)
//    out.writeInt(classifier.parameters.length)
//    for (oneparams <- classifier.parameters) {
//      out.writeInt(oneparams.length)
//      for (param <- oneparams) {
//        out.writeDouble(param)
//      }
//    }
//  }

// For Vim, so we get 4-space tabs
// vim: set ts=4 sw=4 noet:
